h2/src/main/org/h2/server/pg/PgServerThread.java

```diff
@@ -734,6 +739,66 @@ private void writeDataColumn(Value v, int pgType, boolean text) throws IOExcepti
         }
     }
 
+    private static final int[] POWERS10 = {1, 10, 100, 1000, 10000};
+    private static final int MAX_GROUP_SCALE = 4;
+    private static final int MAX_GROUP_SIZE = POWERS10[4];
+
+    private static int divide(BigInteger[] unscaled, int divisor) {
+        BigInteger[] bi = unscaled[0].divideAndRemainder(BigInteger.valueOf(divisor));
+        unscaled[0] = bi[0];
+        return bi[1].intValue();
+    }
+
+    // https://www.npgsql.org/dev/types.html
+    // https://github.com/npgsql/npgsql/blob/8a479081f707784b5040747b23102c3d6371b9d3/
+    //         src/Npgsql/TypeHandlers/NumericHandlers/NumericHandler.cs#L166
+    private void writeNumericBinary(BigDecimal value) throws IOException {
+        int weight = 0;
+        List<Integer> groups = new ArrayList<>();
+        int scale = value.scale();
+        int signum = value.signum();
+        if (signum != 0) {
+            BigInteger[] unscaled = {null};
+            if (scale < 0) {
+                unscaled[0] = value.setScale(0).unscaledValue();
+                scale = 0;
+            } else {
+                unscaled[0] = value.unscaledValue();
+            }
+            if (signum < 0) {
+                unscaled[0] = unscaled[0].negate();
+            }
+            weight = -scale / MAX_GROUP_SCALE - 1;
+            int remainder = 0;
+            int scaleChunk = scale % MAX_GROUP_SCALE;
+            if (scaleChunk > 0) {
+                remainder = divide(unscaled, POWERS10[scaleChunk]) * POWERS10[MAX_GROUP_SCALE - scaleChunk];
+                if (remainder != 0) {
+                    weight--;
+                }
+            }
+            if (remainder == 0) {
+                while ((remainder = divide(unscaled, MAX_GROUP_SIZE)) == 0) {
+                    weight++;
+                }
+            }
+            groups.add(remainder);
+            while (unscaled[0].signum() != 0) {
+                groups.add(divide(unscaled, MAX_GROUP_SIZE));
+            }
+        }
+        int groupCount = groups.size();
+        writeInt(8 + groupCount * 2);
+        writeShort(groupCount);
+        writeShort(groupCount + weight);
+        writeShort(signum < 0 ? 16384 : 0);
+        assert scale >= 0;
+        writeShort(scale > Short.MAX_VALUE ? Short.MAX_VALUE : scale);
```

Comment:
I don't understand this condition. When scale is too large this code simply writes an incorrect scale? Why?

It looks like it should throw an exception instead.

Reply:
Ok. Just throw "Numeric Value out of Range" (22003).